Added support for setting transaction_log_retention_days, retained_backups, and retention_unit to backup_configuration.

tjokimie opened the pull request:
Add support for `transaction_log_retention_days` and `backup_retention_settings`.

Closes https://github.com/terraform-google-modules/terraform-google-sql-db/issues/199.

morgante commented:
Please check the contributing guide and ensure tests pass: https://github.com/terraform-google-modules/terraform-google-sql-db/blob/master/CONTRIBUTING.md

schirayu commented:
@tjokimie - are you still working on this PR?

waqas105 commented:
Anyupdate on this request? i tried the code not wokring for me.

morgante commented:
@waqas105 The PR needs to be rebased and edited to make sure tests pass.

comment-bot-dev commented:
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_documentation
 The documentation needs to be regenerated. Please run `make generate_docs`. 
 ```bash 
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/modules/mssql/README.md /tmp/tmp.p3GJYits9K/generate_docs/workspace/modules/mssql/README.md
17c17
< | backup\_configuration | The database backup configuration. | <pre>object({<br>    binary_log_enabled             = bool<br>    enabled                        = bool<br>    point_in_time_recovery_enabled = bool<br>    start_time                     = string<br>  })</pre> | <pre>{<br>  "binary_log_enabled": null,<br>  "enabled": false,<br>  "point_in_time_recovery_enabled": null,<br>  "start_time": null<br>}</pre> | no |
---
> | backup\_configuration | The database backup configuration. | <pre>object({<br>    binary_log_enabled             = bool<br>    enabled                        = bool<br>    point_in_time_recovery_enabled = bool<br>    start_time                     = string<br>    transaction_log_retention_days = string<br>    backup_retention_settings      = map(string)<br>  })</pre> | <pre>{<br>  "backup_retention_settings": null,<br>  "binary_log_enabled": null,<br>  "enabled": false,<br>  "point_in_time_recovery_enabled": null,<br>  "start_time": null,<br>  "transaction_log_retention_days": null<br>}</pre> | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/modules/mysql/README.md /tmp/tmp.p3GJYits9K/generate_docs/workspace/modules/mysql/README.md
15c15
< | backup\_configuration | The backup\_configuration settings subblock for the database setings | <pre>object({<br>    binary_log_enabled = bool<br>    enabled            = bool<br>    start_time         = string<br>    location           = string<br>  })</pre> | <pre>{<br>  "binary_log_enabled": false,<br>  "enabled": false,<br>  "location": null,<br>  "start_time": null<br>}</pre> | no |
---
> | backup\_configuration | The backup\_configuration settings subblock for the database setings | <pre>object({<br>    binary_log_enabled             = bool<br>    enabled                        = bool<br>    start_time                     = string<br>    location                       = string<br>    transaction_log_retention_days = string<br>    backup_retention_settings      = map(string)<br>  })</pre> | <pre>{<br>  "backup_retention_settings": null,<br>  "binary_log_enabled": false,<br>  "enabled": false,<br>  "location": null,<br>  "start_time": null,<br>  "transaction_log_retention_days": null<br>}</pre> | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/modules/postgresql/README.md /tmp/tmp.p3GJYits9K/generate_docs/workspace/modules/postgresql/README.md
14c14
< | backup\_configuration | The backup\_configuration settings subblock for the database setings | <pre>object({<br>    enabled                        = bool<br>    start_time                     = string<br>    location                       = string<br>    point_in_time_recovery_enabled = bool<br>  })</pre> | <pre>{<br>  "enabled": false,<br>  "location": null,<br>  "point_in_time_recovery_enabled": false,<br>  "start_time": null<br>}</pre> | no |
---
> | backup\_configuration | The backup\_configuration settings subblock for the database setings | <pre>object({<br>    enabled                        = bool<br>    start_time                     = string<br>    location                       = string<br>    point_in_time_recovery_enabled = bool<br>    transaction_log_retention_days = string<br>    backup_retention_settings      = map(string)<br>  })</pre> | <pre>{<br>  "backup_retention_settings": null,<br>  "enabled": false,<br>  "location": null,<br>  "point_in_time_recovery_enabled": false,<br>  "start_time": null,<br>  "transaction_log_retention_days": null<br>}</pre> | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/modules/safer_mysql/README.md /tmp/tmp.p3GJYits9K/generate_docs/workspace/modules/safer_mysql/README.md
173c173
< | backup\_configuration | The backup\_configuration settings subblock for the database setings | <pre>object({<br>    binary_log_enabled = bool<br>    enabled            = bool<br>    start_time         = string<br>    location           = string<br>  })</pre> | <pre>{<br>  "binary_log_enabled": false,<br>  "enabled": false,<br>  "location": null,<br>  "start_time": null<br>}</pre> | no |
---
> | backup\_configuration | The backup\_configuration settings subblock for the database setings | <pre>object({<br>    binary_log_enabled             = bool<br>    enabled                        = bool<br>    start_time                     = string<br>    location                       = string<br>    transaction_log_retention_days = string<br>    backup_retention_settings      = map(string)<br>  })</pre> | <pre>{<br>  "backup_retention_settings": null,<br>  "binary_log_enabled": false,<br>  "enabled": false,<br>  "location": null,<br>  "start_time": null,<br>  "transaction_log_retention_days": null<br>}</pre> | no |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
Running terraform validate
terraform_validate ./examples/mssql-public 
Success!
 The configuration is valid.
terraform_validate ./examples/mysql-ha 
Error: 
Failed to install provider
Error while installing hashicorp/google-beta v3.71.0: unsuccessful request to
https://releases.hashicorp.com/terraform-provider-google-beta/3.71.0/terraform-provider-google-beta_3.71.0_linux_amd64.zip:
500 Service Not Found
terraform_validate ./examples/mysql-private 
Success!
 The configuration is valid.
terraform_validate ./examples/mysql-public 
Success!
 The configuration is valid.
terraform_validate ./examples/postgresql-ha 
Success!
 The configuration is valid.
terraform_validate ./examples/postgresql-public 
Success!
 The configuration is valid.
terraform_validate ./examples/postgresql-public-iam 
Success!
 The configuration is valid.
terraform_validate ./modules/mssql 
Success!
 The configuration is valid.
terraform_validate ./modules/mysql 
Error: 
Failed to install provider
Error while installing hashicorp/google-beta v3.71.0: could not query provider
registry for registry.terraform.io/hashicorp/google-beta: failed to retrieve
authentication checksums for provider: the request failed after 2 attempts,
please try again later: 500
terraform_validate ./modules/postgresql 
Success!
 The configuration is valid.
terraform_validate ./modules/private_service_access 
Success!
 The configuration is valid.
terraform_validate ./modules/safer_mysql 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/mssql-ha 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/mssql-public 
Error: 
Failed to install provider
Error while installing hashicorp/google v3.71.0: unsuccessful request to
https://releases.hashicorp.com/terraform-provider-google/3.71.0/terraform-provider-google_3.71.0_linux_amd64.zip:
500 Service Not Found
terraform_validate ./test/fixtures/mysql-ha 
Error: 
Failed to install provider
Error while installing hashicorp/random v2.3.1: could not query provider
registry for registry.terraform.io/hashicorp/random: failed to retrieve
authentication checksums for provider: the request failed after 2 attempts,
please try again later: 500
Error: 
Failed to install provider
Error while installing hashicorp/null v2.1.2: unsuccessful request to
https://releases.hashicorp.com/terraform-provider-null/2.1.2/terraform-provider-null_2.1.2_linux_amd64.zip:
500 Service Not Found
terraform_validate ./test/fixtures/mysql-private 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/mysql-public 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/postgresql-ha 
Error: 
Failed to install provider
Error while installing hashicorp/null v2.1.2: could not query provider
registry for registry.terraform.io/hashicorp/null: failed to retrieve
authentication checksums for provider: the request failed after 2 attempts,
please try again later: 500
terraform_validate ./test/fixtures/postgresql-public 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/postgresql-public-iam 
Success!
 The configuration is valid.
terraform_validate ./test/setup 
Error: 
Failed to install provider
Error while installing hashicorp/google v3.71.0: could not query provider
registry for registry.terraform.io/hashicorp/google: failed to retrieve
authentication checksums for provider: the request failed after 2 attempts,
please try again later: 500
 ```

tjokimie commented:
> @waqas105 The PR needs to be rebased and edited to make sure tests pass.

@morgante  Sorry, finishing this took quite a bit longer than I anticipated. I fixed the issue, rebased the PR, updated the examples and added some tests.

Some of the tests seem to have some network issues, e.g: 

```
Error while installing hashicorp/null v2.1.2: unsuccessful request to
https://releases.hashicorp.com/terraform-provider-null/2.1.2/terraform-provider-null_2.1.2_linux_amd64.zip:
500 Service Not Found
```

In addition I run the tests locally but I couldn't figure out what triggered this one to fail:

```
×  Postgres SQL pubic instance has just one assigned IP address

     expected: 1
          got: 2

     (compared using ==)
```

The `ipAddresses` block seems to be `[{"ipAddress"=>"34.134.242.179", "type"=>"PRIMARY"}, {"ipAddress"=>"35.239.168.247", "type"=>"OUTGOING"}]`. Do you have any ideas what could cause this?

morgante commented:
/gcbrun

morgante commented on examples/postgresql-ha/main.tf:
```diff
@@ -81,6 +81,11 @@ module "pg" {
     location                       = null
     point_in_time_recovery_enabled = false
+    transaction_log_retention_days = null
+    backup_retention_settings = {
```
I don't think we need to nest this. Let's just collapse to the `backup_configuration` level:

```
backup_configuration = {
    enabled                        = true
    start_time                     = "20:55"
    location                       = null
    point_in_time_recovery_enabled = false
    transaction_log_retention_days = null
    retained_backups = 365
    retention_unit = "COUNT"
}

tjokimie replied:
@morgante Good idea, updated in 8676fba 👍

morgante reviewed (changes requested):
Thanks for the rebase and apologies for the churn. You can ignore the test failure as it looks unrelated.

tjokimie commented:
Rebased with `master` again.

morgante reviewed (approved):
Thanks, test failure is unrelated.
